log_db/vicuna-v1-7b/database_tuning_record.json

```diff
@@ -20,3 +20,4 @@
 [21,[[[["GetBlock",[],["T_reshape","main"],["b0"]],["GetBlock",[],["T_squeeze","main"],["b1"]],["ReverseComputeInline",["b1"],[],[]],["GetLoops",["b0"],[],["l2","l3","l4","l5"]],["Fuse",["l2","l3","l4","l5"],[1],["l6"]],["SampleCategorical",[],[[32,64,128,256],[0.25,0.25,0.25,0.25]],["v7"]],["Split",["l6","None","v7"],[1],["l8","l9"]],["Bind",["l8"],["blockIdx.x"],[]],["Bind",["l9"],["threadIdx.x"],[]],["EnterPostproc",[],[],[]]],[[5,0]]],[2.749852647183396938e-05],{"host":{"keys":["arm_cpu","cpu"],"kind":"llvm","mcpu":"apple-latest","mtriple":"arm64-apple-macos","tag":""},"keys":["metal","gpu"],"kind":"metal","max_function_args":31,"max_num_threads":256,"max_shared_memory_per_block":32768,"max_threads_per_block":256,"tag":"","thread_warp_size":32},[["TENSOR","float16",[1,1,4096]],["TENSOR","float16",[1,32,128]]]]]
+[22,[[[["GetBlock",[],["T_softmax_maxelem","main"],["b0"]],["GetBlock",[],["T_softmax_exp","main"],["b1"]],["GetBlock",[],["T_softmax_expsum","main"],["b2"]],["GetBlock",[],["T_softmax_norm","main"],["b3"]],["GetBlock",[],["root","main"],["b4"]],["ComputeInline",["b1"],[],[]],["GetConsumers",["b2"],[],["b5"]],["GetLoops",["b5"],[],["l6","l7","l8"]],["SampleCategorical",[],[[4,8,16,32,64,128,256,512],[0.125,0.125,0.125,0.125,0.125,0.125,0.125,0.125]],["v9"]],["Split",["l8","None","v9"],[1],["l10","l11"]],["Bind",["l11"],["threadIdx.x"],[]],["ComputeAt",["b2","l7"],[1,-1],[]],["SetScope",["b2"],[0,"shared"],[]],["GetLoops",["b2"],[],["l12","l13","l14","l15","l16"]],["Split",["l16","None","v9"],[1],["l17","l18"]],["Bind",["l18"],["threadIdx.x"],[]],["GetConsumers",["b0"],[],["b19","b20"]],["GetLoops",["b19"],[],["l21","l22","l23","l24","l25","l26"]],["ComputeAt",["b0","l22"],[1,-1],[]],["SetScope",["b0"],[0,"shared"],[]],["GetLoops",["b0"],[],["l27","l28","l29","l30","l31"]],["Split",["l31","None","v9"],[1],["l32","l33"]],["Bind",["l33"],["threadIdx.x"],[]],["SampleCategorical",[],[[0,16,64,512,1024],[0.2000000000000000111,0.2000000000000000111,0.2000000000000000111,0.2000000000000000111,0.2000000000000000111]],["v34"]],["Annotate",["b4","v34"],["meta_schedule.unroll_explicit"],[]],["GetLoops",["b3"],[],["l35","l36","l37","l38"]],["Fuse",["l35","l36"],[1],["l39"]],["Bind",["l39"],["blockIdx.x"],[]],["EnterPostproc",[],[],[]],["GetBlock",[],["root","main"],["b40"]],["Unannotate",["b40"],["meta_schedule.unroll_explicit"],[]],["GetChildBlocks",["b40"],[],["b41","b42","b43"]],["GetLoops",["b41"],[],["l44","l45","l46","l47","l48"]],["GetLoops",["b42"],[],["l49","l50","l51","l52","l53"]],["GetLoops",["b43"],[],["l54","l55","l56"]]],[[8,6],[23,0]]],[0.00011958593939393940078],{"host":{"keys":["arm_cpu","cpu"],"kind":"llvm","mtriple":"arm64-linux-android","tag":""},"keys":["opencl","gpu"],"kind":"opencl","max_num_threads":256,"max_shared_memory_per_block":32768,"max_threads_per_block":1024,"tag":"","texture_spatial_limit":16384,"thread_warp_size":1},[["TENSOR","float32",[1,1,32000]],["TENSOR","float32",[1,1,32000]]]]]
```

Comment:
Besides this, the Adreno target is using the same schedule traces as the iOS metal backend to achieve 12.8 tokens/sec encode and 7.0 tokens/sec decode? If so that is an excellent demonstration of the cross platform capability of TVM Unity. Of course there is plenty of performance left on the table, but this is a very compelling first result!

Reply:
We did need to do some customized schedule for Adreno, https://github.com/mlc-ai/mlc-llm/blob/main/mlc_llm/dispatch/dispatch_tir_operator_adreno.py but they can be done in a way that is quite composable, and we find there are future automation opportunities